File: windsurf/cppchatapp/tests/test_server.cpp
/**
 * @file test_server.cpp
 * @brief Unit tests for ChatServer class
 * 
 * Following AIinfo.txt standards for comprehensive testing.
 * Tests standard cases, edge cases, and wrong inputs.
 */

#include <gtest/gtest.h>
#include "../src/include/interfaces.h"
#include "../src/include/message_impl.h"
#include "../src/server/chat_server.h"
#include <thread>
#include <chrono>
#include <memory>

using namespace chat;

class ServerTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Set up test fixtures
        server = std::make_unique<ChatServer>();
    }
    
    void TearDown() override {
        // Clean up
        if (server && server->is_running()) {
            server->stop_server();
        }
    }
    
    std::unique_ptr<ChatServer> server;
};

// Standard Cases
TEST_F(ServerTest, ServerCreation) {
    EXPECT_NE(server, nullptr);
    EXPECT_FALSE(server->is_running());
    EXPECT_EQ(server->get_connected_clients().size(), 0);
}

TEST_F(ServerTest, ServerWithCustomHandler) {
    auto custom_handler = std::make_unique<DefaultMessageHandler>();
    auto custom_server = std::make_unique<ChatServer>(std::move(custom_handler));
    
    EXPECT_NE(custom_server, nullptr);
    EXPECT_FALSE(custom_server->is_running());
}

TEST_F(ServerTest, FactoryFunction) {
    auto factory_server = create_chat_server();
    
    EXPECT_NE(factory_server, nullptr);
    EXPECT_FALSE(factory_server->is_running());
}

// Edge Cases
TEST_F(ServerTest, StartStopServer) {
    // Test starting server
    EXPECT_TRUE(server->start_server("localhost", 8766));
    EXPECT_TRUE(server->is_running());
    
    // Test stopping server
    server->stop_server();
    EXPECT_FALSE(server->is_running());
}

TEST_F(ServerTest, BroadcastMessageWhileRunning) {
    server->start_server("localhost", 8767);
    
    Message msg("user1", "Hello everyone");
    EXPECT_TRUE(server->broadcast_message(msg));
    
    server->stop_server();
}

TEST_F(ServerTest, BroadcastMessageWhileNotRunning) {
    Message msg("user1", "Hello everyone");
    EXPECT_FALSE(server->broadcast_message(msg));
}

// Wrong Inputs
TEST_F(ServerTest, InvalidPort) {
    // Test with invalid port (should fail gracefully)
    EXPECT_FALSE(server->start_server("localhost", -1));
    EXPECT_FALSE(server->is_running());
}

TEST_F(ServerTest, InvalidHost) {
    // Test with invalid host (should fail gracefully)
    EXPECT_FALSE(server->start_server("invalid_host", 8765));
    EXPECT_FALSE(server->is_running());
}

TEST_F(ServerTest, BroadcastInvalidMessage) {
    server->start_server("localhost", 8768);
    
    Message invalid_msg("", "");  // Invalid message
    EXPECT_FALSE(server->broadcast_message(invalid_msg));
    
    server->stop_server();
}

// Client Management Tests
TEST_F(ServerTest, ClientManagement) {
    server->start_server("localhost", 8769);
    
    // Initially no clients
    EXPECT_EQ(server->get_connected_clients().size(), 0);
    
    server->stop_server();
}

// Message Handler Tests
class MockMessageHandler : public IMessageHandler {
public:
    bool handle_message(const Message& message) override {
        processed_messages.push_back(message);
        return true;
    }
    
    bool validate_message(const Message& message) override {
        validated_messages.push_back(message);
        return message.is_valid();
    }
    
    std::vector<Message> processed_messages;
    std::vector<Message> validated_messages;
};

TEST_F(ServerTest, CustomMessageHandler) {
    auto mock_handler = std::make_unique<MockMessageHandler>();
    auto test_server = std::make_unique<ChatServer>(std::move(mock_handler));
    
    test_server->start_server("localhost", 8770);
    
    Message msg("user1", "Test message");
    EXPECT_TRUE(test_server->broadcast_message(msg));
    
    test_server->stop_server();
}

// Concurrent Tests
TEST_F(ServerTest, ConcurrentClientSimulation) {
    server->start_server("localhost", 8771);
    
    const int num_threads = 5;
    std::vector<std::thread> threads;
    std::vector<Message> messages;
    std::mutex messages_mutex;
    
    // Simulate concurrent message broadcasting
    auto broadcast_messages = [&](int thread_id) {
        for (int i = 0; i < 10; ++i) {
            Message msg(f"user_{thread_id}", f"Message {i} from thread {thread_id}");
            
            if (server->broadcast_message(msg)) {
                std::lock_guard<std::mutex> lock(messages_mutex);
                messages.push_back(msg);
            }
            
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }
    };
    
    // Create threads
    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back(broadcast_messages, i);
    }
    
    // Wait for all threads
    for (auto& thread : threads) {
        thread.join();
    }
    
    server->stop_server();
    
    // Verify messages were processed
    EXPECT_GT(messages.size(), 0);
}

// Performance Tests
TEST_F(ServerTest, MessageBroadcastPerformance) {
    server->start_server("localhost", 8772);
    
    const int num_messages = 1000;
    auto start = std::chrono::high_resolution_clock::now();
    
    for (int i = 0; i < num_messages; ++i) {
        Message msg("user1", f"Message {i}");
        server->broadcast_message(msg);
    }
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    
    server->stop_server();
    
    // Should complete within reasonable time
    EXPECT_LT(duration.count(), 1000000);  // Less than 1 second
    std::cout << "Broadcast performance: " << duration.count() << " microseconds for " 
              << num_messages << " messages" << std::endl;
}

// Error Handling Tests
TEST_F(ServerTest, MultipleStartAttempts) {
    // Start server
    EXPECT_TRUE(server->start_server("localhost", 8773));
    EXPECT_TRUE(server->is_running());
    
    // Try to start again (should fail gracefully)
    EXPECT_FALSE(server->start_server("localhost", 8774));
    
    server->stop_server();
}

TEST_F(ServerTest, MultipleStopAttempts) {
    // Start and stop server normally
    server->start_server("localhost", 8775);
    server->stop_server();
    
    // Try to stop again (should be safe)
    server->stop_server();
    EXPECT_FALSE(server->is_running());
}

// Integration Tests
TEST_F(ServerTest, MessageValidationIntegration) {
    server->start_server("localhost", 8776);
    
    // Test valid message
    Message valid_msg("user1", "Valid message");
    EXPECT_TRUE(server->broadcast_message(valid_msg));
    
    // Test invalid message
    Message invalid_msg("", "Invalid message");
    EXPECT_FALSE(server->broadcast_message(invalid_msg));
    
    server->stop_server();
}

TEST_F(ServerTest, MultipleMessageTypes) {
    server->start_server("localhost", 8777);
    
    std::vector<std::string> message_types = {"text", "image", "file", "system", "error"};
    
    for (const auto& type : message_types) {
        Message msg("user1", f"Content for {type}", type);
        EXPECT_TRUE(server->broadcast_message(msg));
    }
    
    server->stop_server();
}

// Resource Management Tests
TEST_F(ServerTest, ResourceCleanup) {
    {
        auto temp_server = std::make_unique<ChatServer>();
        temp_server->start_server("localhost", 8778);
        temp_server->stop_server();
        // temp_server goes out of scope here
    }
    
    // Should not crash and resources should be cleaned up
    SUCCEED();
}

// Connection Manager Tests
TEST_F(ServerTest, ConnectionManagerIntegration) {
    auto custom_manager = std::make_unique<SocketConnectionManager>();
    auto test_server = std::make_unique<ChatServer>(
        nullptr,  // Use default message handler
        std::move(custom_manager)
    );
    
    EXPECT_TRUE(test_server->start_server("localhost", 8779));
    EXPECT_TRUE(test_server->is_running());
    
    test_server->stop_server();
}

// Stress Tests
TEST_F(ServerTest, StressTest) {
    server->start_server("localhost", 8780);
    
    const int num_threads = 10;
    const int messages_per_thread = 100;
    std::vector<std::thread> threads;
    
    auto stress_broadcast = [&](int thread_id) {
        for (int i = 0; i < messages_per_thread; ++i) {
            Message msg(f"user_{thread_id}", f"Stress message {i}");
            server->broadcast_message(msg);
        }
    };
    
    // Create stress threads
    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back(stress_broadcast, i);
    }
    
    // Wait for completion
    for (auto& thread : threads) {
        thread.join();
    }
    
    server->stop_server();
    
    // Should complete without crashing
    SUCCEED();
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: windsurf/pythonchatapp/tests/test_server.py
"""
Unit tests for chat server implementation.
Following AIinfo.txt standards for comprehensive testing.
"""

import unittest
import asyncio
import json
import websockets
from unittest.mock import Mock, patch, AsyncMock
import sys
import os

# Add src to path for imports
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from server.chat_server import ChatServer, MessageHandler, create_chat_server
from interfaces import Message


class TestMessageHandler(unittest.TestCase):
    """Test cases for MessageHandler class."""
    
    def setUp(self):
        """Set up test fixtures."""
        self.handler = MessageHandler()
    
    def test_message_handling_standard_case(self):
        """Test standard message handling."""
        message = Message("user1", "Hello world")
        
        # Test validation
        result = asyncio.run(self.handler.validate_message(message))
        self.assertTrue(result)
        
        # Test handling (should not raise exception)
        asyncio.run(self.handler.handle_message(message))
    
    def test_message_handling_edge_cases(self):
        """Test edge cases for message handling."""
        # Long message
        long_message = Message("user", "a" * 1000)
        result = asyncio.run(self.handler.validate_message(long_message))
        self.assertTrue(result)
        
        # Special characters
        special_message = Message("user", "Special chars: !@#$%^&*()_+-=[]{}|;':\",./<>?")
        result = asyncio.run(self.handler.validate_message(special_message))
        self.assertTrue(result)
    
    def test_message_handling_wrong_inputs(self):
        """Test wrong inputs for message handling."""
        # Empty message
        empty_message = Message("", "")
        result = asyncio.run(self.handler.validate_message(empty_message))
        self.assertFalse(result)
        
        # Too long message
        too_long_message = Message("user", "a" * 1001)
        result = asyncio.run(self.handler.validate_message(too_long_message))
        self.assertFalse(result)


class TestChatServer(unittest.TestCase):
    """Test cases for ChatServer class."""
    
    def setUp(self):
        """Set up test fixtures."""
        self.server = ChatServer()
    
    def tearDown(self):
        """Clean up after tests."""
        if self.server.is_running:
            asyncio.run(self.server.stop_server())
    
    def test_server_creation(self):
        """Test server creation and initial state."""
        self.assertFalse(self.server.is_running)
        self.assertEqual(len(self.server.get_connected_clients()), 0)
    
    def test_server_with_custom_handler(self):
        """Test server creation with custom message handler."""
        custom_handler = MessageHandler()
        server = ChatServer(custom_handler)
        
        self.assertEqual(server.message_handler, custom_handler)
        self.assertFalse(server.is_running)
    
    def test_server_factory_function(self):
        """Test factory function for server creation."""
        server = create_chat_server()
        
        self.assertIsInstance(server, ChatServer)
        self.assertFalse(server.is_running)
    
    @patch('websockets.serve')
    def test_server_start_stop(self, mock_serve):
        """Test server start and stop functionality."""
        # Mock the websocket server
        mock_server_instance = Mock()
        mock_serve.return_value = mock_server_instance
        
        # Test starting server
        result = asyncio.run(self.server.start_server("localhost", 8765))
        self.assertTrue(result)
        self.assertTrue(self.server.is_running)
        
        # Test stopping server
        asyncio.run(self.server.stop_server())
        self.assertFalse(self.server.is_running)
    
    def test_broadcast_message_standard_case(self):
        """Test standard message broadcasting."""
        message = Message("user1", "Hello everyone")
        
        # Mock the server as running
        self.server.is_running = True
        
        # Test broadcasting (should not raise exception)
        result = asyncio.run(self.server.broadcast_message(message))
        self.assertTrue(result)
    
    def test_broadcast_message_edge_cases(self):
        """Test edge cases for message broadcasting."""
        # Server not running
        self.server.is_running = False
        message = Message("user1", "Hello")
        result = asyncio.run(self.server.broadcast_message(message))
        self.assertFalse(result)
        
        # Invalid message
        self.server.is_running = True
        invalid_message = Message("", "")
        result = asyncio.run(self.server.broadcast_message(invalid_message))
        self.assertFalse(result)
    
    def test_client_management(self):
        """Test client connection management."""
        # Initially no clients
        clients = self.server.get_connected_clients()
        self.assertEqual(len(clients), 0)
        
        # Mock client connections
        self.server.clients = {
            "client1": Mock(),
            "client2": Mock()
        }
        
        clients = self.server.get_connected_clients()
        self.assertEqual(len(clients), 2)
        self.assertIn("client1", clients)
        self.assertIn("client2", clients)


class TestChatServerIntegration(unittest.TestCase):
    """Integration tests for chat server."""
    
    def setUp(self):
        """Set up test fixtures."""
        self.server = ChatServer()
    
    def tearDown(self):
        """Clean up after tests."""
        if self.server.is_running:
            asyncio.run(self.server.stop_server())
    
    def test_message_validation_flow(self):
        """Test complete message validation and broadcasting flow."""
        # Set up server as running
        self.server.is_running = True
        
        # Create valid message
        message = Message("user1", "Test message")
        
        # Test validation and broadcasting
        result = asyncio.run(self.server.broadcast_message(message))
        self.assertTrue(result)
    
    def test_multiple_message_types(self):
        """Test handling of different message types."""
        self.server.is_running = True
        
        message_types = ["text", "image", "file", "system"]
        
        for msg_type in message_types:
            message = Message("user", f"Content for {msg_type}", msg_type)
            result = asyncio.run(self.server.broadcast_message(message))
            self.assertTrue(result)
    
    def test_concurrent_client_simulation(self):
        """Test simulation of concurrent client connections."""
        import threading
        import time
        
        # Mock client connections
        clients = {}
        lock = threading.Lock()
        
        def add_client(client_id):
            with lock:
                clients[client_id] = Mock()
                time.sleep(0.01)  # Simulate connection time
        
        # Add multiple clients concurrently
        threads = []
        for i in range(10):
            thread = threading.Thread(target=add_client, args=(f"client_{i}",))
            threads.append(thread)
            thread.start()
        
        # Wait for all threads
        for thread in threads:
            thread.join()
        
        # Verify all clients were added
        self.assertEqual(len(clients), 10)
        
        # Update server clients
        self.server.clients = clients
        connected_clients = self.server.get_connected_clients()
        self.assertEqual(len(connected_clients), 10)


class TestServerErrorHandling(unittest.TestCase):
    """Test cases for server error handling."""
    
    def setUp(self):
        """Set up test fixtures."""
        self.server = ChatServer()
    
    def test_invalid_port_handling(self):
        """Test handling of invalid port numbers."""
        with patch('websockets.serve', side_effect=Exception("Invalid port")):
            result = asyncio.run(self.server.start_server("localhost", -1))
            self.assertFalse(result)
    
    def test_network_error_handling(self):
        """Test handling of network errors."""
        with patch('websockets.serve', side_effect=ConnectionError("Network error")):
            result = asyncio.run(self.server.start_server("localhost", 8765))
            self.assertFalse(result)
    
    def test_message_broadcast_errors(self):
        """Test error handling during message broadcasting."""
        self.server.is_running = True
        
        # Mock client that raises exception during send
        mock_client = Mock()
        mock_client.send = Mock(side_effect=Exception("Connection lost"))
        self.server.clients = {"client1": mock_client}
        
        message = Message("user1", "Test message")
        
        # Should handle error gracefully
        result = asyncio.run(self.server.broadcast_message(message))
        self.assertTrue(result)  # Still returns True as error is handled internally


if __name__ == '__main__':
    # Configure test runner
    unittest.main(verbosity=2)
